transaction/core/tests/input_rules.rs

```diff
@@ -63,25 +72,290 @@ fn test_input_rules_verify_max_tombstone() {
+            .get_value_from_amount_shared_secret(
+                &r_txo.amount_shared_secret[..].try_into().unwrap()
+            )
+            .unwrap()
+            .0
+    );
+
+    result.tx_out.masked_amount = Some(MaskedAmount::V2(new_masked_amount));
+    result
+}
+
+// Test that input rules verification is working for fractional output rules
+#[test]
+fn test_input_rules_verify_fractional_outputs() {
+    let block_version = BlockVersion::THREE;
+
+    let mut rng: RngType = SeedableRng::from_seed([7u8; 32]);
+    let alice = AccountKey::random(&mut rng);
+    let alice_pub = alice.default_subaddress();
+    // Amount is 4000, so we will create 4 tx outs worth 1000
+    let (mut tx, _ledger) =
+        create_test_tx_with_amount_and_comparer_and_recipients::<DefaultTxOutputsOrdering>(
+            block_version,
+            4000,
+            INITIALIZE_LEDGER_AMOUNT - 4000,
+            &[&alice_pub, &alice_pub, &alice_pub, &alice_pub],
+        );
+
+    // Modify the Tx to have some (empty) input rules.
+    // (This invalidates the signature, but we aren't checking that here)
+    tx.prefix.inputs[0].input_rules = Some(InputRules::default());
+
+    // Check that the Tx is following input rules (vacuously)
+    get_first_rules(&tx).verify(block_version, &tx).unwrap();
+
+    // We're going to pull the four TxOut's from the TxPrefix, and make
+    // corresponding fractional outputs and fractional change from them, also
+    // rewriting their amounts when we do that, to see what happens to the
+    // validation routine. In order to do that, we need to get their amount
+    // shared secrets, and since Alice is the owner, we need Alice's
+    // account keys to do that.
+    assert_eq!(tx.prefix.outputs.len(), 4);
+    assert!(tx.prefix.outputs[0].public_key != tx.prefix.outputs[1].public_key);
+    let revealed_tx_outs: Vec<RevealedTxOut> = tx
+        .prefix
+        .outputs
+        .iter()
+        .map(|txo| {
+            let decompressed_tx_pub = RistrettoPublic::try_from(&txo.public_key).unwrap();
+            let tx_out_shared_secret =
+                get_tx_out_shared_secret(alice.view_private_key(), &decompressed_tx_pub);
+            let amount_shared_secret =
+                MaskedAmount::compute_amount_shared_secret(block_version, &tx_out_shared_secret)
+                    .unwrap();
+            let (amount, _) = txo
+                .masked_amount
+                .as_ref()
+                .unwrap()
+                .get_value_from_amount_shared_secret(&amount_shared_secret)
+                .unwrap();
+            assert_eq!(amount.value, 1000);
+            assert_eq!(*amount.token_id, 0);
+            RevealedTxOut {
+                tx_out: txo.clone(),
+                amount_shared_secret: amount_shared_secret.to_vec(),
+            }
+        })
+        .collect();
+
+    // Add a fractional output, by doubling one of the revealed tx outs.
+    // So this means we are filling the 1/2 of the request.
```

Comment:
Starting at 0.5 is a less valuable test, but it's also easier for the reviewer to follow as a warmup. We have subsequent tests that do 2/3 and 1/3.

Reply:
> I think the underlying logic is incorrect.

Can you explain what precisely you think is incorrect?

Reply:
> I'm going to put the numbers back to [MICP 42[(https://github.com/garbageslam/mcips/blob/partial-fill-rules/text/0042-partial-fill-rules.md)

If you mainly want to discuss the example from the MCIP, that MCIP is also open for review, and we could discuss there rather than in this PR review

Reply:
>  : where does Alices 980 MOB "change" get accounted for? I think this is part of my hiccup on the "fractional change" term. I think of 980 being the "change" but it may be my lack of financial domain knowledge here.

It gets accounted for as the "real change output" corresponding to the "fractional change output" which appears in the input rules.

>  It was mentioned we aren't verifying the signatures so maybe that's the piece, but how do we know the 200MEOWB output is for Alice?

We can't verify this cryptographically, but we can assume that Alice is directing profits from the trade to some account that she controls. To some extent, whenever Alice is able to direct a location to which money flows, Alice "controls" that money.

>  In the example from the MCIP what prevents Bob from dropping off the 500MOB change to Alice?

This is exactly the same mechanism as what prevents Bob from dropping a required output for Alice in MCIP 31. Bob cannot tamper with the InputRules or it invalidate's Alice's MLSAG signature which signed away her money in the first place. If the InputRules are left undisturbed, Bob must supply a "real output" that corresponds to each "fractional output", and the code in this PR is called by the enclave to validate that that happens. So Bob's only choice is to provide sufficient fractional change, or the Tx will be invalid.

Reply:
The other thing I will say is, we added a number of tests in `transaction/std/src/signed_contingent_input_builder.rs` that test all this functionality at a higher level (using the SCI builder and the Transaction builder), and those tests all seem to be working as expected. But if we got the low-level arithmetic relating to fractional change wrong, then all those tests should be failing.

So that leaves me doubtful that "the underlying logic is incorrect", but if you see a specific issue let's talk about it.

Reply:
I'm going to assume after this comment:

> Getting here I think I finally understand that txout_0 is Alice's "change" from the money she fronted

that you actually don't believe the underlying logic is incorrect now, and what seems wrong is the terminology.